### PR TITLE
Fire/Air alarms properly dismantle when they are destroyed & fixes balloon alert runtimes when cutting terminals

### LINKED ISSUE #39756: /obj/machinery/power/apc/cant_delet
Literally can't destroy or delete it.

image

image

Either i am a total imbecile or this APC can't be fully constructed (i.e. power areas)

`while i was writing this issue the cargo APC finally disappeared and so did some others what i tried to delete, but not all. Doesn't seem like server lag, people weren't complaining and i could smash stuff with an energy axe`

### PR DESCRIPTION
**Pull request by SyncIt21**
Firealarms & Airalarms `deconstruct()` procs are wrong, they spew out electronics & cable regardless of what build stage their on

Fire alarms providing free electronics & cable without even installing them

https://user-images.githubusercontent.com/110812394/233784785-650fbd64-3c6c-44c8-b377-c3a8244d28ac.mp4

Air alarms providing free electronics & cable without even installing them

https://user-images.githubusercontent.com/110812394/233784799-cdd38967-8a59-454d-8da4-1360d03ff12c.mp4

Bug exploits now come to an end

Also fixes balloon alerts run timing when cutting terminals connected to apc's/smes etc. because the atom gets deleted, by the time the balloon alert gets executed it adds a timer on the deleted object causing runtimes.

## Changelog
:cl:
fix: air alarms correctly spew out cable & electronics depending on their build stage when their integrity reaches 0  i.e. destroyed
fix: fire alarms correctly spew out cable & electronics depending on their build stage when their integrity reaches 0  i.e. destroyed
fix: balloon alerts runtiming when cutting terminals
/:cl:

### CODE REVIEW

**Comment by zxaber** [comment]
It's somewhat nice that they don't revert to metal automatically, because you can locate the broken APC quickly and know the reason a room has no power. If the APC was fully destroyed and you don't know the location of a given room's APC by heart, you may end up searching for a while before realizing it's gone.

**Comment by SyncIt21** [comment]
> It's somewhat nice that they don't revert to metal automatically, because you can locate the broken APC quickly and know the reason a room has no power. If the APC was fully destroyed and you don't know the location of a given room's APC by heart, you may end up searching for a while before realizing it's gone.

#39756 clearly reported this behaviour as a bug, anything that reaches 0 integrity must disappear, you cannot justify a bug as a convenience. In the event a big bomb, singularity, whatever destroys the whole room/station we want no traces remaining and so a single floating apc in the middle of nowhere would be weird. 
Yes this is a degrade for quality of life but correct behaviour takes higher priority.

**Comment by LemonInTheDark** [comment]
https://github.com/tgstation/tgstation/issues/39756 was wrong, apc shells being left over as a sign of the destruction is intentional behavior (people are bad at reconstructing them, it's important we don't make that process harder)
I don't like it, but I consider it better then the alternative

**Comment by SyncIt21** [comment]
> #39756 was wrong, apc shells being left over as a sign of the destruction is intentional behavior (people are bad at reconstructing them, it's important we don't make that process harder) I don't like it, but I consider it better then the alternative

LOL sure I'll revert it. It's dumb and stupid that an apc can survive even an singularity and still be just floating there

**Comment by SyncIt21** [comment]
bu

**Comment by SyncIt21** [comment]
Wrong button

**Comment by SyncIt21** [comment]
Ok apc changes are reverted.

File: code/modules/atmospherics/machinery/air_alarm/air_alarm_interact.dm [hunk 1]
```diff
@@ -92,10 +92,12 @@
...
 	if(!(flags_1 & NODECONSTRUCT_1))
 		new /obj/item/stack/sheet/iron(loc, 2)
-		var/obj/item/I = new /obj/item/electronics/airalarm(loc)
-		if(!disassembled)
-			I.take_damage(I.max_integrity * 0.5, sound_effect=FALSE)
-		new /obj/item/stack/cable_coil(loc, 3)
+		if((buildstage & AIR_ALARM_BUILD_NO_WIRES) || (buildstage & AIR_ALARM_BUILD_COMPLETE))
+			var/obj/item/I = new /obj/item/electronics/airalarm(loc)
+			if(!disassembled)
+				I.take_damage(I.max_integrity * 0.5, sound_effect = FALSE)
```
Comment (san7890) [inline code/modules/atmospherics/machinery/air_alarm/air_alarm_interact.dm:98]:
```suggestion
			var/obj/item/alarm = new /obj/item/electronics/airalarm(loc)
			if(!disassembled)
				alarm.take_damage(alarm.max_integrity * 0.5, sound_effect = FALSE)
```

may as well clean up this single-letter variable while in the area

**Review by san7890 — CHANGES_REQUESTED** [review]

File: code/modules/atmospherics/machinery/air_alarm/air_alarm_interact.dm (hunk 1 above)
Comment (san7890) [reply code/modules/atmospherics/machinery/air_alarm/air_alarm_interact.dm:98]: also we could also do something like

```dm
var/obj/item/electronics/airalarm/alarm = new(loc)
```

but that's just shifting the words to a different spot in the line and isn't _truly_ necessary

File: code/game/machinery/firealarm.dm [hunk 2]
```diff
@@ -421,7 +421,7 @@
 	. = ..()
 	if(.) //damage received
 		if(atom_integrity > 0 && !(machine_stat & BROKEN) && buildstage != 0)
-			if(prob(33))
+			if(prob(33) && buildstage == 2) //require fully wired electronics to set of the alarms
```
Comment (ZephyrTFA) [inline code/game/machinery/firealarm.dm:424]: magic numbers :(

File: code/modules/atmospherics/machinery/air_alarm/air_alarm_interact.dm [hunk 3]
```diff
@@ -92,10 +92,12 @@
 /obj/machinery/airalarm/deconstruct(disassembled = TRUE)
 	if(!(flags_1 & NODECONSTRUCT_1))
 		new /obj/item/stack/sheet/iron(loc, 2)
-		var/obj/item/I = new /obj/item/electronics/airalarm(loc)
-		if(!disassembled)
-			I.take_damage(I.max_integrity * 0.5, sound_effect=FALSE)
-		new /obj/item/stack/cable_coil(loc, 3)
+		if((buildstage & AIR_ALARM_BUILD_NO_WIRES) || (buildstage & AIR_ALARM_BUILD_COMPLETE))
```
Comment (ZephyrTFA) [inline code/modules/atmospherics/machinery/air_alarm/air_alarm_interact.dm:95]: Why are you bit comparing here. Surely these aren't bit flags

**Review by ZephyrTFA — CHANGES_REQUESTED** [review]

**Comment by SyncIt21** [comment]
Done

**Pull request merged**